src/ray/object_manager/pull_manager.cc

```diff
@@ -39,33 +42,183 @@ uint64_t PullManager::Pull(const std::vector<rpc::ObjectReference> &object_ref_b
     it->second.bundle_request_ids.insert(bundle_it->first);
   }
 
+  // We have a new request. Activate the new request, if the
+  // current available memory allows it.
+  UpdatePullsBasedOnAvailableMemory(num_bytes_available_);
+
   return bundle_it->first;
 }
 
+bool PullManager::ActivateNextPullBundleRequest(
+    std::map<uint64_t, std::vector<rpc::ObjectReference>>::iterator next_request_it,
+    std::unordered_set<ObjectID> *object_ids_to_pull) {
+  // Check that we have sizes for all of the objects in the bundle. If not, we
+  // should not activate the bundle, since it may put us over the available
+  // capacity.
+  for (const auto &ref : next_request_it->second) {
+    auto obj_id = ObjectRefToId(ref);
+    const auto it = object_pull_requests_.find(obj_id);
+    RAY_CHECK(it != object_pull_requests_.end());
+    if (!it->second.object_size_set) {
+      RAY_LOG(DEBUG) << "No size for " << obj_id << ", canceling activation for pull "
+                     << next_request_it->first;
+      return false;
+    }
+  }
+
+  // Activate the bundle.
+  for (const auto &ref : next_request_it->second) {
+    auto obj_id = ObjectRefToId(ref);
+    bool start_pull = active_object_pull_requests_.count(obj_id) == 0;
+    active_object_pull_requests_[obj_id].insert(next_request_it->first);
+    if (start_pull) {
+      RAY_LOG(DEBUG) << "Activating pull for object " << obj_id;
+      // This is the first bundle request in the queue to require this object.
+      // Add the size to the number of bytes being pulled.
+      // NOTE(swang): The size could be 0 if we haven't received size
```

Comment:
If size is not known (0), is it not reaching this line due to 63? Is there anything I am missing?

Reply:
Ah oops, this comment was for the other block of code. Will move it.